js/ui/components/clubhouse.js

```diff
@@ -0,0 +1,323 @@
+                                     layout_manager: new Clutter.BinLayout() });
+
+        Main.layoutManager.addChrome(this.actor);
+
+        this.source = new ClubhouseNotificationSource(CLUBHOUSE_ID);
+        this.source.connect('notify', Lang.bind(this, this._onNotify));
+
+        this._overrideAddNotification();
+    },
+
+    disable: function() {
+        if (!this._useClubhouse)
+            return;
+
+        log('Disabling Clubhouse component');
+
+        if (this.actor)
+            this.actor.destroy();
+
+        if (this._clubhouseProxyHandler > 0)
+            this._clubhouseProxy.disconnect(this._clubhouseProxyHandler);
+
+        if (this._oldAddNotificationFuncPrototype) {
+            NotificationDaemon.GtkNotificationDaemon.prototype.AddNotificationAsync =
+                this._oldAddNotificationFuncPrototype;
+        }
+
+        this.actor = null;
+        this.source = null;
+        this._banner = null;
+        this._clubhouseProxy = null;
+    },
+
+    _imageUsesClubhouse: function(eosImageVersion) {
+        // We are only using the Clubhouse component in Endless Hack images for now, so
+        // check if the prefix of the image version matches the mentioned flavor.
+        return eosImageVersion.startsWith('hack-');
+    },
+
+    _overrideAddNotification: function() {
+        this._oldAddNotificationFuncPrototype =
+            NotificationDaemon.GtkNotificationDaemon.prototype.AddNotificationAsync;
+        NotificationDaemon.GtkNotificationDaemon.prototype.ClubhouseData = [CLUBHOUSE_ID,
+                                                                            this.source];
+        NotificationDaemon.GtkNotificationDaemon.prototype.AddNotificationAsync = function(params, invocation) {
+            let [appId, notificationId, notification] = params;
+
+            let clubhouseId = this.ClubhouseData[0];
+
+            // If the app sending the notification is the Clubhouse, then use our own source
+            if (appId == clubhouseId) {
+                let source = this.ClubhouseData[1];
+                source.addNotification(notificationId, notification, true);
+                invocation.return_value(null);
+                return;
+            }
+
+            oldFuncPrototype.apply(Main.notificationDaemon._gtkNotificationDaemon, arguments);
+        }
+    },
+
+    _onNotify: function(source, notification) {
+        this.actor.visible = true;
+
+        // Only create the banner if it doesn't yet exist, otherwise just rely on the notification
+        // object getting updated.
+        if (!this._banner) {
+            this._banner = notification.createBanner();
+            this._bannerCloseHandler = this._banner.connect('dismiss',
+                                                            Lang.bind(this, this._dismiss_quest_cb))
+            this.actor.add_child(this._banner.actor);
+            this._reposition();
+        }
+    },
+
+    _dismiss_quest_cb: function() {
```

Comment:
Typically camel case is used as naming convention for functions in the shell.

Reply:
Arrgh. I did fix a bunch of snake-case vars but apparently I missed this one. Thanks.